Crafting Button Enhancements

13spacemen opened the pull request:
NEW:


https://user-images.githubusercontent.com/46101244/183987813-f30dd772-368c-4e9e-9dbf-d3e447fcfa2b.mp4




OLD:

https://user-images.githubusercontent.com/46101244/180233793-e3a508b7-9994-439f-a592-6b50a251f751.mp4

![finalformcrafting](https://user-images.githubusercontent.com/46101244/180329672-06f05a94-71b5-41e6-8b78-89c204d8c33c.png)

<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
The T-shaped crafting button becomes a lightbulb if the items in your hands are used in crafting recipes

Right clicking the crafting button will show all possible recipes for items in user's hands
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
So players don't have to memorize huge lists of crafting recipes.
<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
qol: The crafting button lights up if you can craft something with an item in hands. Epic!
qol: Right-clicking the crafting button will show all possible recipes for items in your hands. Double epic!
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

13spacemen committed:
added GLOB.crafting_recipe_items for items used in crafting recipes
added GLOB.crafting_recipe_items_blacklist for blacklisted crafting items
crafting button becomes a lightbulb if current held item is used in recipes

Jacquerel commented:
Does it light up like that if you can currently craft something or just if the item you are holding is used in some recipe somewhere, regardless of whether the surrounding material is sufficient to make it?  
I wouldn't want this to light up literally every time I am holding a saw, battery, syringe, or various other very common tool items.

13spacemen commented:
@Jacquerel it lights up if the item you're holding is used in some recipe somewhere
the crafting recipe craftable check searches through your bag and items on turfs next to you too
not sure how to implement changing the crafting button icon for that

dragomagol commented:
I like the idea behind it but I'm not super into the the implementation. If I'm playing the game and picking things up I don't really want a lightbulb flashing in the corner of my eye to remind me that because I'm holding some iron and rods I can make a skateboard.

LemonInTheDark commented on code/datums/components/crafting/recipes.dm:
```diff
@@ -41,6 +41,18 @@
+		var/actual_item = new item(null)
+		if(!isitem(actual_item))
+			continue
+		if(!(item in GLOB.crafting_recipe_items))
+			GLOB.crafting_recipe_items += item
```
bro what

13spacemen replied:
The item might already be in the GLOB list, since it's a for loop and it might have been added while looping through a previous recipe

LemonInTheDark commented on code/modules/mob/living/carbon/human/human_update_icons.dm:
```diff
@@ -600,8 +600,17 @@ There are several things that need to be remembered:
+
 	var/list/hands = list()
 	for(var/obj/item/worn_item in held_items)
+		if(((worn_item.type in GLOB.crafting_recipe_items) || ((worn_item.parent_type in GLOB.crafting_recipe_items) && !(worn_item.type in GLOB.crafting_recipe_items_blacklist))) && hud_used)
```
bro just make a typecache it's what you want here anyway

13spacemen replied:
I'm not sure what a typecache is or how to do so, could you give me a few pointers?

13spacemen committed:
adds RMB screentip to crafting button
right clicking crafting button will display possible recipes for items in user's hands

13spacemen committed:
typecaches
results have their own list
fixed tool_paths name

13spacemen committed:
multi-hand support

LemonInTheDark commented on code/datums/components/crafting/recipes.dm:
```diff
@@ -41,6 +41,18 @@
+		if(!isitem(actual_item))
+			continue
+		if(!(item in GLOB.crafting_recipe_items))
+			GLOB.crafting_recipe_items += item
```
1. you do not need to new() actual_item, just use ispath()
2. similarly, rather then checking if the item is already in the list of types, just use |=

LemonInTheDark commented on code/modules/mob/living/carbon/human/human_update_icons.dm:
```diff
@@ -600,8 +600,17 @@ There are several things that need to be remembered:
+
 	var/list/hands = list()
 	for(var/obj/item/worn_item in held_items)
+		if(((worn_item.type in GLOB.crafting_recipe_items) || ((worn_item.parent_type in GLOB.crafting_recipe_items) && !(worn_item.type in GLOB.crafting_recipe_items_blacklist))) && hud_used)
```
These should be typecaches.
the `in` keyword requires iterating the whole list, whereas typecaches use associated lists, which are logarithmic time, can' remember which

LemonInTheDark commented on code/modules/mob/living/carbon/human/human_update_icons.dm:
```diff
@@ -600,8 +600,17 @@ There are several things that need to be remembered:
+		if(((worn_item.type in GLOB.crafting_recipe_items) || ((worn_item.parent_type in GLOB.crafting_recipe_items) && !(worn_item.type in GLOB.crafting_recipe_items_blacklist))) && hud_used)
+			for(var/atom/movable/screen/craft/crafting_button in hud_used.static_inventory)
+				crafting_button.icon = 'icons/hud/screen_gen.dmi'
+				crafting_button.icon_state = "can_craft"
```
this is not the proper way to change the icon, since any hud resets would break the effect. this needs to be done differently

LemonInTheDark commented on code/datums/components/crafting/recipes.dm:
```diff
@@ -41,6 +41,20 @@
 		tool_behaviors = string_list(tool_behaviors)
 	if(tool_paths)
 		tool_paths = string_list(tool_paths)
+	if(ispath(result, /obj/item))
+		GLOB.crafting_recipe_items_results |= result
+		GLOB.crafting_recipe_items_results[result] = TRUE
+	for(var/item in blacklist)
+		if(!ispath(item, /obj/item) || item == result)
+			continue
+		if(!is_type_in_typecache(item, GLOB.crafting_recipe_items))
+			GLOB.crafting_recipe_items_blacklist |= item
+			GLOB.crafting_recipe_items_blacklist[item] = TRUE
+	for(var/item in reqs)
+		if(!ispath(item, /obj/item))
+			continue
+		GLOB.crafting_recipe_items |= item
+		GLOB.crafting_recipe_items[item] = TRUE
```
you don't need |= here it's ok

LemonInTheDark reviewed (changes requested):
yo based I reviewed the wrong commit
still think the icon thing smells off but the typecache stuff is better

optimumtact commented:
This has broken something about rendering given the screenshot test failures

Mothblocks commented:
@optimumtact From what I can tell that's old:

![image](https://user-images.githubusercontent.com/35135081/182979988-0993a845-67aa-4731-835f-13a76b743edc.png)

This would be :x: if it was still happening. Should probably auto-hide old results though.

optimumtact commented:
ohhh

13spacemen committed:
removes unnecessary |=

13spacemen committed:
added reset icon proc
crafting button only changes icon if picking up an item from a turf, resets with a timer

13spacemen committed:
made icon change an overlay instead, more visually cohesive

13spacemen committed:
use of signals

13spacemen committed:
add documentation
alter overlays updating method

LemonInTheDark commented on code/_onclick/hud/screen_objects.dm:
```diff
@@ -89,6 +89,94 @@
+
+/atom/movable/screen/craft/Click(location, control, params)
+	var/list/modifiers = params2list(params)
+	if(LAZYACCESS(modifiers, RIGHT_CLICK))
+		for(var/hand_number in 0 to length(hud.mymob.held_items))
+			var/item_in_hand = hud.mymob.get_item_for_held_index(hand_number)
+			display_possible_recipes(item_in_hand)
+		if(length(hud.mymob.held_items) == length(hud.mymob.get_empty_held_indexes()))
+			to_chat(hud.mymob, span_warning("You aren't holding anything in your hands!"))
+	else
+		..()
```
```suggestion
	if(!LAZYACCESS(modifiers, RIGHT_CLICK))
		return ..()

		for(var/hand_number in 0 to length(hud.mymob.held_items))
		var/item_in_hand = hud.mymob.get_item_for_held_index(hand_number)
		display_possible_recipes(item_in_hand)
	if(length(hud.mymob.held_items) == length(hud.mymob.get_empty_held_indexes()))
		to_chat(hud.mymob, span_warning("You aren't holding anything in your hands!"))	
```
early returns to make this easier to read

LemonInTheDark commented on code/_onclick/hud/screen_objects.dm:
```diff
@@ -89,6 +89,94 @@
+	var/recipe_line = ""
+	var/all_recipes = ""
+	for(var/datum/crafting_recipe/recipe in GLOB.crafting_recipes)
+		if((item.type in recipe.reqs) || ((item.parent_type in recipe.reqs) && !(item.type in recipe.blacklist)) || item.type == recipe.result)
+			possible_recipes += recipe
```
why do we count it if this is a potential result?

LemonInTheDark commented on code/_onclick/hud/screen_objects.dm:
```diff
@@ -89,6 +89,94 @@
+			for(var/tool in recipe.tool_behaviors)
+				tools_text += tool
+		recipe_line = "<b>[recipe.name]</b>: [jointext(reqs_text, ", ")][(length(reqs_text) && length(tools_other_text)) ? ", " : null][span_green("[jointext(tools_other_text, ", ")][((length(tools_other_text) || length(reqs_text)) && length(tools_text))  ? ", " : null][jointext(tools_text, ", ")]")]"
+		all_recipes += "[recipe_line]\n"
```
rather then doing it like this, store all_recipes as a list, add recipe_line directly to it, then when you're done turn it into a text string with .Join("\n")

This is both faster, and easier to parse

LemonInTheDark commented on code/_onclick/hud/screen_objects.dm:
```diff
@@ -89,6 +89,94 @@
+		if(length(recipe.tool_behaviors))
+			for(var/tool in recipe.tool_behaviors)
+				tools_text += tool
+		recipe_line = "<b>[recipe.name]</b>: [jointext(reqs_text, ", ")][(length(reqs_text) && length(tools_other_text)) ? ", " : null][span_green("[jointext(tools_other_text, ", ")][((length(tools_other_text) || length(reqs_text)) && length(tools_text))  ? ", " : null][jointext(tools_text, ", ")]")]"
```
what is this line doing and why is it so long

LemonInTheDark commented on code/_onclick/hud/screen_objects.dm:
```diff
@@ -89,6 +89,94 @@
+	else
+		..()
+
+/atom/movable/screen/craft/proc/display_possible_recipes(obj/item/item)
```
please document the purpose of this proc

LemonInTheDark commented on code/_onclick/hud/screen_objects.dm:
```diff
@@ -89,6 +89,94 @@
+/atom/movable/screen/craft/update_overlays()
+	. = ..()
+	if(has_craft_items)
+		. += image('icons/hud/screen_gen.dmi', icon_state = "can_craft")
```
could you use mutable_appearance for this? it has the same effect, but it's consistent with how we handle overlays

LemonInTheDark commented on code/_onclick/hud/screen_objects.dm:
```diff
@@ -89,6 +89,94 @@
+	SIGNAL_HANDLER
+	if(slot != ITEM_SLOT_HANDS)
+		for(var/obj/item/hand_item in user.held_items)
+			if(hand_item.crafting_item_in_hands && hand_item != item)
```
so you're using `.crafting_item_in_hands` to cache that typecache lookup. smart

LemonInTheDark replied:
rather then storing that information on the item though, why don't we store a list of weakrefs on the crafting button?
That way everything related to us is stored on us, instead of somewhere else

LemonInTheDark commented on code/_onclick/hud/screen_objects.dm:
```diff
@@ -89,6 +89,94 @@
+			item.crafting_item_in_hands = FALSE
+			update_appearance()
+			return
+	if(is_type_in_typecache(item, GLOB.crafting_recipe_items) || (is_type_in_typecache(item.parent_type, GLOB.crafting_recipe_items) && !is_type_in_typecache(item.type, GLOB.crafting_recipe_items_blacklist)) || is_type_in_typecache(item, GLOB.crafting_recipe_items_results))
```
could you split this up with early returns so it's easier to read?

LemonInTheDark commented on code/_onclick/hud/screen_objects.dm:
```diff
@@ -89,6 +89,94 @@
+
+/atom/movable/screen/craft/proc/exited_check(mob/living/carbon/human/user, obj/item/item, direction)
+	SIGNAL_HANDLER
+	for(var/obj/item/hand_item in user.held_items)
+		if(hand_item.crafting_item_in_hands && hand_item != item)
+			return
```
also it saves needing to walk held_items, and actually replaces has_craft_items.
would just be a length(list) check

LemonInTheDark commented on code/_globalvars/lists/objects.dm:
```diff
@@ -25,6 +25,9 @@ GLOBAL_LIST(fake_reagent_blacklist) //List of all reagents that are parent types
 GLOBAL_LIST_EMPTY(tech_list) //list of all /datum/tech datums indexed by id.
 GLOBAL_LIST_EMPTY(surgeries_list) //list of all surgeries by name, associated with their path.
 GLOBAL_LIST_EMPTY(crafting_recipes) //list of all table craft recipes
+GLOBAL_LIST_EMPTY(crafting_recipe_items) //list of all crafting recipe requirement /obj/item
+GLOBAL_LIST_EMPTY(crafting_recipe_items_blacklist) //list of blacklisted crafting recipe /obj/item
+GLOBAL_LIST_EMPTY(crafting_recipe_items_results) //list of crafting recipe result /obj/item
```
do you mean recipes here?

13spacemen committed:
wipppp

13spacemen commented:
taking a break for a while

LemonInTheDark commented:
it was almost done 
![northern_ice](https://user-images.githubusercontent.com/58055496/184509887-d71249ad-3276-4e42-95e1-71122e8e3016.jpg)

optimumtact commented:
they are conflicted
